collect_app/src/main/java/org/odk/collect/android/fragments/dialogs/CustomDatePickerDialog.java

```diff
@@ -0,0 +1,195 @@
+                    }
+                })
+                .setNegativeButton(R.string.cancel, new DialogInterface.OnClickListener() {
+                    @Override
+                    public void onClick(DialogInterface dialog, int id) {
+                        dismiss();
+                    }
+                })
+                .create();
+    }
+
+    @Override
+    public void onResume() {
+        super.onResume();
+        gregorianDateText = (TextView) getDialog().findViewById(R.id.date_gregorian);
+        setUpPickers();
+    }
+
+    @Override
+    public void onSaveInstanceState(Bundle outState) {
+        outState.putSerializable(FORM_INDEX, formIndex);
+        outState.putSerializable(DATE, getDateAsGregorian(getOriginalDate()));
+        outState.putSerializable(DATE_PICKER_DETAILS, datePickerDetails);
+
+        super.onSaveInstanceState(outState);
+    }
+
+    private void setUpPickers() {
+        dayPicker = (NumberPicker) getDialog().findViewById(R.id.day_picker);
+        dayPicker.setOnValueChangedListener(new NumberPicker.OnValueChangeListener() {
+            @Override
+            public void onValueChange(NumberPicker picker, int oldVal, int newVal) {
+                updateGregorianDateLabel();
+            }
+        });
+        monthPicker = (NumberPicker) getDialog().findViewById(R.id.month_picker);
+        monthPicker.setOnValueChangedListener(new NumberPicker.OnValueChangeListener() {
+            @Override
+            public void onValueChange(NumberPicker picker, int oldVal, int newVal) {
+                updateDays();
+                updateGregorianDateLabel();
+            }
+        });
+        yearPicker = (NumberPicker) getDialog().findViewById(R.id.year_picker);
+        yearPicker.setOnValueChangedListener(new NumberPicker.OnValueChangeListener() {
+            @Override
+            public void onValueChange(NumberPicker picker, int oldVal, int newVal) {
+                updateDays();
+                updateGregorianDateLabel();
+            }
+        });
+
+        hidePickersIfNeeded();
+    }
+
+    private void hidePickersIfNeeded() {
+        if (datePickerDetails.isMonthYearMode()) {
+            dayPicker.setVisibility(View.GONE);
+            dayPicker.setValue(1);
+        } else if (datePickerDetails.isYearMode()) {
+            dayPicker.setVisibility(View.GONE);
+            monthPicker.setVisibility(View.GONE);
+            dayPicker.setValue(1);
+            yearPicker.setValue(1);
+        }
+    }
+
+    private LocalDateTime getDateAsGregorian(LocalDateTime date) {
+        return DateTimeUtils.skipDaylightSavingGapIfExists(date)
+                .toDateTime()
+                .withChronology(GregorianChronology.getInstance())
+                .toLocalDateTime();
+    }
+
+    protected static Bundle getArgs(FormIndex formIndex, LocalDateTime date, DatePickerDetails datePickerDetails) {
+        Bundle args = new Bundle();
+        args.putSerializable(FORM_INDEX, formIndex);
```

Comment:
We should be careful about serializing the JavaRosa classes just because some of them feature some very deep nesting and we could end up serializing a lot more data than we need to. I think its *okay* here, as FormIndex looks simple enough, but its something we should watch out for.

Personally, I would add an `onPickerClick` method to the listener, and keep the `FormIndex` out of this class entirely, allowing the Widget to handle its own call to `setIndexWaitingForData`.

Reply:
I think it's ok. FormIndex is used in the same way in other places as well as I remember and there is also a test: https://github.com/opendatakit/javarosa/commit/87fc3dcf69b55fa9548284e4a88e612527ae71a0#diff-00bf7c2c715f66829ca3844e22334112

Reply:
@grzesiek2010 is there a specific reason not to follow what @jknightco is recommending with `onPickerClick`? Perhaps you two could quickly discuss in #collect-code so we're making a conscious decision here. 😊 In general I agree that keeping `FormIndex` out would be a good thing.